### ISSUE TITLE
🦠 Model Request: ImageMol Embeddings

### ISSUE THREAD

**Issue by DhanshreeA** [ersilia-os/ersilia#567]
### Model Name

Molecular representation learning

### Model Description

Representation Learning Framework that utilizes molecule images for encoding molecular inputs as machine readable vectors for downstream tasks such as bio-activity prediction, drug metabolism analysis, or drug toxicity prediction. The approach utilizes transfer learning, that is, pre-training the model on massive unlabeled datasets to help it in generalizing feature extraction and then fine tuning on specific tasks.

### Slug

image-mol-embeddings

### Tag

Embedding

### Publication

https://www.nature.com/articles/s42256-022-00557-6

### Source code

https://github.com/HongxinXiang/ImageMol

### License

MIT License

**Comment by DhanshreeA** [comment]
Related to #518

**Comment by DhanshreeA** [comment]
Updates:

I got the embeddings for ImageMol by loading the pretrained model provided by the authors, and then creating a Sequential block out of the model's children except the last fully connected classification layer. 

```python
model_blocks = list(model.children())
print(f"Total blocks in model: {len(model_blocks)}")
model_embeddings = torch.nn.Sequential(*model_blocks[:-1]) # Remove the final classifier head
print(f"Total blocks in embeddings: {len(list(model_embeddings.children()))}")

Total blocks in model: 10
Total blocks in embeddings: 9
```

To further validate whether these embeddings are actually meaningful, as suggested by Miquel and Gemma, I carried out some experiments on similar and dissimilar molecules. 

**Experimental setup**

I took the following molecule (input_mol): `Nc1nc(NC2CC2)c2ncn([C@H]3C=C[C@@H](CO)C3)c2n1` (abacavir, anti-viral administered for HIV) and generated molecules similar to it using the Similarity search in ChEMBL, DrugBank and UNPD from the Ersilia model hub  (https://github.com/ersilia-os/eos9c7k)
And for dissimilar molecules, I took 100 SMILES from [eml_canonical.csv](https://github.com/ersilia-os/ersilia/files/10565788/eml_canonical.csv), since this list has significant overlap with FDA approved drugs that may not be similar to the molecule under consideration.

For both the similar and dissimilar molecules, I generated their MorganFingerprints and ImageMol embeddings. I used the input_mol's MorganFP and ran tanimoto similarity between it and every molecule in the similar and dissimilar sets.
Then I generated ImageMol embeddings for the input_mol as well as every molecule in the similar and dissimilar sets and calculated distance metrics for measuring similarity, namely, Euclidean and Manhattan distances.
Furthermore, since the scale of the Tanimoto score and distance metrics vary widely sine Tanimoto scores lie betwee [0,1], and distance scores lie between [0, N), I normalized distance scores using min-max normalization to lie between the [0,1] range such that trends, if any, are clearly visible.

Attached below are the results:

Dissimilar Molecules:
![dissimilar](https://user-images.githubusercontent.com/7213136/216250755-dbf0e335-a639-4cef-883f-14f9e689bf05.png)

Similar Molecules:
![similar](https://user-images.githubusercontent.com/7213136/216250764-c50fd595-fbf6-4f5d-b6ea-02f8afa98fdb.png)

Both of these similarity measures operate differently from each other, ie, Tanimoto score increases directly with increasing similarity and vice-versa; whereas distance measures decrease with increasing similarity and vice-versa. Therefore the trends I expected to see in these plots are:

- Similar case: high Tanimoto, low Manhattan and Euclidean
- Dissimilar case: low Tanimoto, high Manhattan and Euclidean.

This does appear to be the case, ie for the Similar case, the Tanimoto score distribution roughly lies above the Manhattan and Euclidean score distributions; whereas for the Dissimilar case, the Tanimoto distribution lies roughly below the Manhattan and Euclidean distributions.

**Comment by DhanshreeA** [comment]
PR: https://github.com/ersilia-os/eos4avb/pull/1

**Comment by DhanshreeA** [comment]
Currently resolving: 
https://github.com/ersilia-os/eos4avb/issues/3
https://github.com/ersilia-os/eos4avb/issues/4

**Comment by DhanshreeA** [comment]
@GemmaTuron should we close this issue?
